### ISSUE TITLE
Feature request: partial data

### ISSUE THREAD

**Issue by jantorestolsvik** [apollographql/apollo-feature-requests#104]
Migrated from: apollographql/apollo-client#2425

**Comment by levrik** [comment]
I would also really appreciate a feature like this.
My situation is the following. I'm fetching a list of documents with file name, status and some other fields. When you click on it a more detailed view opens. Which displays the file name and status in the header. The rest of the data has to be fetched then. I would like to be able to re-use the file name and status for the header while the rest is loading. I tried setting `returnPartialData` to `true` but this doesn't give the desired result since all these fields are defined as non-nullable in the schema.
Now I have to pass the information that should be shown in the header to my component which itself fetches the additional information of the same document which feels quite wrong.

**Comment by seeruk** [comment]
Yeah, I have a similar use-case here. I'm showing breadcrumbs on every page in my app. The name of the first element in the breadcrumbs is based on a slug in the URL which is resolved as part of the query that runs on every page. I've essentially got a specific query with a specific fragment in it that is always run on each page to ensure the data is there - but of course because the data is grabbed every page load the first time if the whole query isn't in the cache, I have to show a loading indicator between pages (and they could move moving between pages by interacting with these breadcrumbs, so that'd be particularly jarring).

`returnPartialData` does work, but when using TypeScript, the generated types are basically lies because then any of the top-level queries can be null. I'm not entirely sure what the solution is, maybe some way of batching up 2 "separate" queries into one request, then splitting the result back out so you can cache separately, but only query once?

**Comment by smeijer** [comment]
My case is a bit simpler but still requires `partial data`.

I believe `readFragment` and `writeFragment` are often combined to update the cache. The throwing of `readFragment` when one or more fields are missing, makes it harder to deal with cache updates. Having a `partial` flag here would be most welcome.

```js
const fragment = gql`
  fragment BlogComments on Blog {
    _id
    comments {
      body
    }
  }
`;

const newComment = { body: 'hi' };

const blog = store.readFragment({ 
  id: 'blogId', 
  fragment: fragment,
});

const nextBlog = Object.assign({}, blog, {
  comments: [...(blog.comments || []), newComment],
}

store.writeFragment({
  id: 'blogId',
  fragment,
  data: nextBlog,
});
```

If `blog` here does not yet contain the `comments` property in the cache, this will fail due to the error that `readFragment` is throwing. Which is unfortunate. if `readFragment` would return `null` or `Partial<T>`, it would become less cumbersome to deal with this kind of scenario.

**Comment by LydiaF** [comment]
Dear Apollo,

Is there any possibility that this partial flag on readFragment is on the horizon?

Thanks for any info :)

**Comment by dylanwulf** [comment]
This might help: https://github.com/appmotion/apollo-augmented-hooks

**Comment by VadimZhiltsov** [comment]
voting for the feature, as it is important in BFF use case for Apollo Server
